### PR TITLE
ASC-918 Add Support for Test Case Steps

### PR DESCRIPTION
**Pull request by ryan-rs**
Test cases marked with 'test_step' property marked with 'true' in the JUnitXML
will be treated as test steps for a combined test case with test case name
derived from the 'classname' JUnitXML attribute.

This feature required a ton of changes. As I worked through the changes I found
a bunch of small formatting issues that I fixed along the way. The diff is
super gross. Sorry | not sorry?

### CODE REVIEW

File: zigzag/zigzag_test_log.py [hunk 1]
```diff
@@ -4,51 +4,65 @@
...
 
     _TESTCASE_NAME_RGX = re.compile(r'(^[\w-]+)')
     _test_run_failure_output_field_id = 0
     _fields = 0
     _failure_link_field_id = 0
     _test_sha_field_id = 0
 
-    def __init__(self, testcase_xml, mediator):
-        """Create a TestLog object
+    def __init__(self, testcase_xml, mediator, auto_parse=True, auto_lookup=True):
```
Comment (zreichert) [inline zigzag/zigzag_test_log.py]: Do `auto_parse` and `auto_lookup` need to be two separate options?

File: zigzag/zigzag_test_log.py [hunk 2]
```diff
@@ -512,3 +536,190 @@ def _get_truncated_failure_output(cls, message):
...
+
+        # Initialize without parsing so we can post-process the test steps.
+        super(_ZigZagTestLogWithSteps, self).__init__(None, mediator, auto_parse=False, auto_lookup=False)
+
+        self._name = testcase_name
+        self._teststeps_xml = teststeps_xml
+        self._zz_test_step_logs = []
+        self._qtest_test_step_logs = []
+
+        self._parse()
```
Comment (zreichert) [inline zigzag/zigzag_test_log.py]: Looks like you are parsing here even though the signature says `auto_parse=False`.

File: zigzag/zigzag_test_log.py [hunk 3]
```diff
@@ -512,3 +536,190 @@ def _get_truncated_failure_output(cls, message):
...
+
+        Args:
+            testcase_name (str): The name of the test case.
+            teststeps_xml (list(ElementTree)): A list of JUnit style testcase XML elements representing test steps for
+                a single qTest test log.
+            mediator (ZigZag): The mediator that stores shared data.
+        """
+
+        # Initialize without parsing so we can post-process the test steps.
+        super(_ZigZagTestLogWithSteps, self).__init__(None, mediator, auto_parse=False, auto_lookup=False)
```
Comment (zreichert) [inline zigzag/zigzag_test_log.py]: It looks like `auto_parse` and `auto_lookup` are unused.

File: zigzag/zigzag_test_log.py [hunk 4]
```diff
@@ -4,51 +4,65 @@
...
-from zigzag.module_hierarchy_facade import ModuleHierarchyFacade
 from zigzag.link_generation_facade import LinkGenerationFacade
+from zigzag.module_hierarchy_facade import ModuleHierarchyFacade
 
 
-class ZigZagTestLog(object):
+# ======================================================================================================================
+# Classes
+# ======================================================================================================================
+class _ZigZagTestLog(object):
```
Comment (zreichert) [inline zigzag/zigzag_test_log.py]: Can you provide some insight into why you want to make this class private?

**Comment by ryan-rs** [comment]
@zreichert I added integration tests demonstrating requirement linking for tests with steps.

**Pull request merged**